server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,240 @@
+  extends SparkApp
+  with Logging {
+  import SparkYarnApp.ec
+
+  private var state: SparkApp.State = SparkApp.State.STARTING
+  private var yarnDiagnostics: IndexedSeq[String] = IndexedSeq.empty[String]
+
+  override def log(): IndexedSeq[String] =
+    (process map (_.inputLines) getOrElse ArrayBuffer.empty[String]) ++ yarnDiagnostics
+
+  override def kill(): Unit = synchronized {
+    if (isRunning) {
+      process foreach (_.destroy())
+
+      try {
+        Await.result(appIdFuture map (yarnClient.killApplication), SparkYarnApp.KILL_TIMEOUT)
+      } catch {
+        // We cannot kill the YARN app without the app id.
+        // There's a chance the YARN app hasn't been submitted during a livy-server failure.
+        // We don't want a stuck session that can't be deleted. Emit a warning and move on.
+        case _: TimeoutException | _: InterruptedException =>
+          warn("Deleting a session while its YARN application is not found.")
+          yarnAppMonitorThread.interrupt()
+      }
+    }
+  }
+
+  private def changeState(newState: SparkApp.State.Value): Unit = {
+    if (state != newState) {
+      listener foreach (_.stateChanged(state, newState))
+      state = newState
+    }
+  }
+
+  private def getYarnDiagnostics(appReport: ApplicationReport): IndexedSeq[String] = {
+    Option(appReport.getDiagnostics)
+      .filter(_.nonEmpty)
+      .map[IndexedSeq[String]]("YARN Diagnostics:" +: _.split("\n"))
+      .getOrElse(IndexedSeq.empty)
+  }
+
+  private def isRunning: Boolean = {
+    state != SparkApp.State.FAILED && state != SparkApp.State.FINISHED &&
+    state != SparkApp.State.KILLED
+  }
+
+  // Exposed for unit test.
+  private[utils] def mapYarnState(
+      appId: ApplicationId,
+      yarnAppState: YarnApplicationState,
+      finalAppStatus: FinalApplicationStatus): SparkApp.State.Value = {
+    yarnAppState match {
+      case (YarnApplicationState.NEW |
+            YarnApplicationState.NEW_SAVING |
+            YarnApplicationState.SUBMITTED |
+            YarnApplicationState.ACCEPTED) => SparkApp.State.STARTING
+      case YarnApplicationState.RUNNING => SparkApp.State.RUNNING
+      case YarnApplicationState.FINISHED =>
+        finalAppStatus match {
+          case FinalApplicationStatus.SUCCEEDED => SparkApp.State.FINISHED
+          case FinalApplicationStatus.FAILED => SparkApp.State.FAILED
+          case FinalApplicationStatus.KILLED => SparkApp.State.KILLED
+          case s =>
+            error(s"Unknown YARN final status $appId $s")
+            SparkApp.State.FAILED
+        }
+      case YarnApplicationState.FAILED => SparkApp.State.FAILED
+      case YarnApplicationState.KILLED => SparkApp.State.KILLED
+    }
+  }
+
+  // Exposed for unit test.
+  // TODO Instead of spawning a thread for every session, create a centralized thread and
+  // batch YARN queries.
+  private[utils] val yarnAppMonitorThread = new Thread(s"yarnAppMonitorThread-$this") {
```

Comment:
Having a utility method to start a daemon thread would be nice, since you do this in at least a couple of places. Something like:

    def startDaemonThread(name: String)(fn: => Unit): Thread = {
      val thread = new Thread(new Runnable() {
        override def run(): Unit = fn
      }
      ...
    }